### PR TITLE
feat(pipeline-ui): app code view and source editing

### PR DESCRIPTION
**Pull request by tiina303**
## Problem

source code view and editing for source apps

## Changes

👉 *Stay up-to-date with [PostHog coding conventions](https://posthog.com/docs/contribute/coding-conventions) for a smoother review.*

### CODE REVIEW

**Comment by tiina303** [comment]
cc @Twixes or @thmsobrmlr  two n00b questions:
1. is there a way for me to display the code without it being editable (for non source apps)?
2. in the current implementation if I expand multiple rows it gets weird - shows the wrong code in the other one, how can I fix it, or would it be the best to only allow one to be expanded at a time?

**Comment by thmsobrmlr** [comment]
Hey @tiina303! I've added a couple of commits regarding your questions. Happy to pair on this as well if it's easier :) Didn't address anything other than the questions for now.

> It's WIP: @Twixes or @thmsobrmlr two n00b questions:
> 
> 1. is there a way for me to display the code without it being editable (for non source apps)?

Yep, there's a `readOnly` option for the monaco editor that make the code read only.

> 2. in the current implementation if I expand multiple rows it gets weird - shows the wrong code in the other one, how can I fix it, or would it be the best to only allow one to be expanded at a time?

The reason for this is that the monaco editor refers to the same "model". Here's an explanation of how the multi-model feature of our monaco wrapper works: https://www.npmjs.com/package/@monaco-editor/react#multi-model-editor. Concretely adding the pluginId to the file path helped.

---

I've also added an example for a more lightweight code snippet in case editing here isn't strictly necessary.

As I said, let me know if you want to pair or have follow up questions. Great work so far, was happy to see the logics work as intended 🙌

File: frontend/src/scenes/pipeline/AppCode.tsx [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+                        </>
+                    ) : (
+                        <LemonButton type="secondary" onClick={() => setEditMode(true)}>
+                            Edit
+                        </LemonButton>
+                    )}
+                </div>
+            )}
+            <CodeEditor
+                // TODO: For some reason the editor doesn't keep the cursor focus and then I can type only one char and need to click again
```
Comment (tiina303) [inline frontend/src/scenes/pipeline/AppCode.tsx:60]: is there an easy way I can fix this. I saw that in the old UI it's a form and then the onChange is there

Comment (Twixes) [reply frontend/src/scenes/pipeline/AppCode.tsx:60]: There are two possibilities:
1. Something else steals the focus (unlikely)
2. This `CodeEditor` component is rendered with a different identity when its `value` changes. In my experience, this happens, but I don't immediately see why in this case. Might be something with `LemonTabs`? The React Dev Tools' have a tree view which is really handy for tracking down re-renders like this

Comment (tiina303) [reply frontend/src/scenes/pipeline/AppCode.tsx:60]: somewhat related: I'm thinking we only want a single app code edited at once probably - so a pop-up modal for edit and in here it's readOnly?

Comment (tiina303) [reply frontend/src/scenes/pipeline/AppCode.tsx:60]: > The React Dev Tools' have a tree view which is really handy for tracking down re-renders like this

you can show me on Monday :)

**Review by Twixes — APPROVED** [review]

**Pull request merged**